src/polkit/polkitunixprocess.c

```diff
@@ -151,8 +156,11 @@ struct _PolkitUnixProcess
 
   gint pid;
   guint64 start_time;
+  guint64 cgroupid;
   gint uid;
   gint pidfd;
+  gint ppidfd;
```

Comment:
i don't think looking at the ppid is a good idea, because that means subshells (some of which are somewhat hidden in shells) would not qualify anymore.

instead: check the posix session id (i.e. the one managed by setsid()/getsid(), not the audit session id, nor the logind session id, which are three distinct things). the posix session id should be unique for each shell, and is necessary for correct shell sesson mgmt (it means that if the shell goes away all forked of processes get SIGHUP). 

the session leader has pid=sid btw, and if it exits the session becomes orphaned. (thus the sid will exist for longer, until all processes with it actually exited, but now without a leader). but that's good enough, it allows tracking sids via pidfds for this purpose, and thus get a proper pidfd inode id for the sid, that is safe regarding recycling

Reply:
> i don't think looking at the ppid is a good idea, because that means subshells (some of which are somewhat hidden in shells) would not qualify anymore.

That was intentional: right now auth is checked on every single invocation. This PR relaxes this somewhat, by allowing multiple commands in the same shell to be done without checking auth again. Letting scripts running in the same shell also run without auth is a step too far for now. We can think about it again later, but for now I prefer to keep it very explicit, so that if you do:

```
$ pkexec foo
$ pkexec bar
```

You only get auth'ed once, but if you do:

```
$ pkexec foo
$ ./bar.sh
```

and bar.sh itself calls pkexec, auth is checked again. This was it's very very explicit, or at least it tries to be. We can consider further relaxing it more down the line. There are other things to consider in the future too, to match sudo's behaviour: whether to increase the default window to 15m, and whether to "refresh" the token on each invocation. But for now let's keep it nice and tight.

> the posix session id should be unique for each shell

- it _should_ be, but it's not mandatory, it depends on the shell implementation itself
- it is 32bit so it can wrap around, and the only way to make that safe is to use the super-new pidfd filesystem (introduced this year) which is 64bit only and requires a disabled-by-default kconfig to be enabled

so it's riskier, and cuts off basically all existing Linux systems, and might be usable in a year or two, so it's too restrictive, so for now I'll keep checking the tty

Reply:
> it should be, but it's not mandatory, it depends on the shell implementation itself

no it does not? the sid is always initialized from the pid of the caller. you don't get to choose your sid?

ptys are tightly packed (i.e. you always get lowest free pty), and aggressively recycled. otoh, sids are just like pids, i.e. allocated form a counter basically.

in other words, there is no rational rason why you think checking ptys is fine but checking sids is not.

sids change between user sessions on the same ttys, but tty numbers are super stable, hence not suitable for authentication that shall be context bound.

Reply:
> no it does not? the sid is always initialized from the pid of the caller. you don't get to choose your sid?

But that means, for all intents and purposes, sid == pid, including the fact that it wraps around and doesn't require privileges to get a new one, and hit a collision.

> sids change between user sessions on the same ttys, but tty numbers are super stable, hence not suitable for authentication that shall be context bound.

sudo uses ttys to cache